proto/proxy/destination/destination.proto

```diff
@@ -10,12 +10,40 @@ import "common/common.proto";
 // weighted set of addresses and address metadata.  Can be implemented with DNS
 // or lookups against other service discovery backends.
 //
+// If the service does not exist then the controller must send
+// `no_endpoints{exists: false}` ASAP when a client subscribes or when the
+// service stops existing. If the service exists and has endpoints available
+// then the controller must send `add` that lists all (or at least a large
+// number) of the endpoints for the service. If and only if the service exists
+// but does not have any endpoints available then the controller SHOULD send
+// `no_endpoints{exists: true}` when a client subscribes. In other words, the
+// `no_endpoints` message must only be sent when there are *no*endpoints for
+// the service.
+//
 // The controller is expected to send an Update every time there is a
 // change in service discovery.  The controller is also expected to send an
 // update at least once every ADDRESS_UPDATE_INTERVAL to indicate that the
 // controller is still healthy.  If no service discovery updates have taken
 // place, the controller can simply send an empty `add`.  The controller may
 // determine the value of ADDRESS_UPDATE_INTERVAL.
+//
+// The client MUST be prepared to receive messages in any order and the client
+// MUST be able to cope with the presence or absence of redundant messages.
+//
+// `no_endpoints` followed by an `add` is *not* equivalent to just sending the
+// `add` regardless of the value of the `exists` field in the `no_endpoints`
+// message. `remove` followed by a `no_endpoints` message is equivalent to
+// sending just the `no_endpoints` message, and a `remove` that removes the
+// last endpoint is equivalent to a `no_endpoints{exists: true}` message.
```

Comment:
The `NoEndpoints` message contains an explicit indication of whether the service exists. `Remove(...)` would be ambiguous as to whether the service exists, and that distinction is important.

Also, depending on the implementation in the server, it might be useful for the server to use the `NoEndpoints` message instead of `Remove(...)` if it is removing the endpoints for the service due to a change in the service's status observed through the `Services` API instead of the `Endpoints` API in Kubernetes. I'm not sure.

Reply:
I had assumed that `Remove(<all addresses>)` would unambiguously indicate that the service does exist but has no endpoints.  Is that not correct?

Reply:
> I had assumed that Remove(<all addresses>) would unambiguously indicate that the service does exist but has no endpoints. Is that not correct?

Yes, that's right. But one would have to read the documentation to understand what it means at the beginning of a connection before anything has been added. The proxy handles it the way you suggest mostly for completeness's sake.

Definitely this isn't the only way to formulate the API. If you have a specific change you want to make, please suggest it.

Reply:
Just to clarify, my questions are just for my own understanding and should absolutely not block moving forward with this implementation.